### ISSUE TITLE
Support emitting comments, string blocks and alternate integer encodings in serialized output

### ISSUE THREAD

**Issue by cfrantz** [dtolnay/serde-yaml#235]
I'd like to have support for emitting richer output from the serializer.  My motivation is to improve outputs which are meant to be both human and machine readable.  For me, this would include support for:
- Emitting comments
- Emitting multi-line string blocks (e.g. `|-` and `|+` forms).
- Emitting integers in alternate bases (dec, hex, bin, oct).

Informing the serializer about how to produce these alternate forms should be relatively easy.  The user should be able to derive a trait on the struct (e.g. `YamlFormat`) and specify the desired formatting information in attributes (e.g. `#[yaml(format=hex)]`).

## Example
Consider a diagnostic tool which reads the Serial Flash Discoverable Parameters (SFDP) from a SPI eeprom.  I'd like to emit the `signature` field as hex and the `reserved` field as binary and I'd like to emit some comments which explain the data in the struct.

```
#[derive(Serialize, Deserialize, YamlFormat)]
struct SfdpHeader {
    #[yaml(format=hex, comment="Signature should be 0x50444653")]
    signature: u32,
    #[yaml(comment = "SFDP version")]
    minor: u8,
    major: u8,
    #[yaml(comment = "Number of parameter headers minus 1")]
    nph: u8,
    #[yaml(format=bin, comment="Reserved field should be all ones")]
    reserved: u8,
}
```

The serialized form of the above struct might look like this:
```
# Signature should be 0x50444653
signature: 0x50444653
# SFDP version
minor: 6
major: 1
# Number of parameter headers minus 1
nph: 2
# Reserved field should be all ones
reserved: 0b11111111
```

## Sample Implementation
I have created a sample implementation in #234 with support for comments, multi-line string blocks and alternate integer encodings.  There are several samples in [`tests/test_format.rs`](https://github.com/cfrantz/serde-yaml/blob/yamlformat/tests/test_format.rs).

The sample implementation may not be the best approach (it's my first try at this sort of thing).  I'd like get some feedback on both desired features and the approach to the implementation.

**Comment by alevinval** [comment]
Hey, just a lurker. I was interested in supporting comments as well. Not only that but respecting the existing comments when loading the YAML back. The main problem is that the underlying YAML parser is mimicking the canonical C implementation which does not support parsing or emitting them, also the `yaml-rust` crate is not actively maintained anymore. I've seen some open PRs adding support for comments, but it all seemed very hacky or loosely implemented. So I've spent the last 10 days wrapping my head around the parser codebase.

First, I focused on adding support in the scanner to be able to emit `Comment` tokens, that was easy. Second, figure out a way to extend the parser to emit comments, without interfering with the parser state machine and its transitions. Third, figure out a way to update the YAML node loader to handle the comments, and finally updated the emitter. Just a few hours ago I have succeeded in having some basic comment support that does not break any existing tests and correctly round-trips with this [sample yaml file with comments](https://github.com/alevinval/yaml-rust/pull/7/files#diff-687937e1c52b6e0080c70ccc9bdde14a35f5dfdfa6d0105c508b19181cafde6fR567). I do not think it is anywhere close to being finalized, but today I have seen some light at the end of the tunnel.

I just wanted to give a heads up to you folks (specially @dtolnay) in case you find this work interesting, and maybe even useful for the long-term plans of the `serde-yaml` library. In the meanwhile, I will continue development to solidify the `yaml-rust` parser to have robust comment support.
